### ISSUE TITLE
Telemetry fails to restart

### ISSUE THREAD

**Issue by costinm** [istio/istio#7050]
This morning my cluster had nothing in grafana, traced it back to telemetry job.

Logs from mixer:

```
 Mixer started with
MaxMessageSize: 1048576
MaxConcurrentStreams: 1024
APIWorkerPoolSize: 1024
AdapterWorkerPoolSize: 1024
APIPort: 9091
APIAddress: unix:///sock/mixer.socket
MonitoringPort: 9093
EnableProfiling: true
SingleThreaded: false
NumCheckCacheEntries: 1500000
ConfigStoreURL: k8s://
ConfigDefaultNamespace: istio-system
LoggingOptions: log.Options{OutputPaths:[]string{"stdout"}, ErrorOutputPaths:[]string{"stderr"}, RotateOutputPath:"", RotationMaxSize:104857600, RotationMaxAge:30, RotationMaxBackups:1000, JSONEncoding:false, LogGrpc:true, outputLevels:"default:info", logCallers:"", stackTraceLevels:"default:none"}
TracingOptions: tracing.Options{ZipkinURL:"http://zipkin:9411/api/v1/spans", JaegerURL:"", LogTraceSpans:false}
IntrospectionOptions: ctrlz.Options{Port:0x2694, Address:"127.0.0.1"}
Unable to initialize Mixer: unable to listen: listen unix /sock/mixer.socket: bind: address already in use 
```

kublectl describe:

```
 Mixer started with
MaxMessageSize: 1048576
MaxConcurrentStreams: 1024
APIWorkerPoolSize: 1024
AdapterWorkerPoolSize: 1024
APIPort: 9091
APIAddress: unix:///sock/mixer.socket
MonitoringPort: 9093
EnableProfiling: true
SingleThreaded: false
NumCheckCacheEntries: 1500000
ConfigStoreURL: k8s://
ConfigDefaultNamespace: istio-system
LoggingOptions: log.Options{OutputPaths:[]string{"stdout"}, ErrorOutputPaths:[]string{"stderr"}, RotateOutputPath:"", RotationMaxSize:104857600, RotationMaxAge:30, RotationMaxBackups:1000, JSONEncoding:false, LogGrpc:true, outputLevels:"default:info", logCallers:"", stackTraceLevels:"default:none"}
TracingOptions: tracing.Options{ZipkinURL:"http://zipkin:9411/api/v1/spans", JaegerURL:"", LogTraceSpans:false}
IntrospectionOptions: ctrlz.Options{Port:0x2694, Address:"127.0.0.1"}
Unable to initialize Mixer: unable to listen: listen unix /sock/mixer.socket: bind: address already in use 
```

**Comment by costinm** [comment]
Noticed in the console an 'out of memory' error - so it is possible mixer container crashed, and k8s retried to restart the container in same pod and kept failing.

**Comment by kyessenov** [comment]
Thanks for catching this. We should set all istio pods with sidecars to have "restartPolicy=Never".

**Comment by costinm** [comment]
Marking as P0 since it affects stability of the system, if policy server has the same problem it can bring down everything.

**Comment by costinm** [comment]
Well, we should clean up the socket - restartPolicy is not the root cause here.

**Comment by kyessenov** [comment]
There are two issues:
1. OOMing Mixer -- why is that happening?
2. Partial Mixer pod restarts. There is no reason to keep the proxy running if mixer crashes.

**Comment by kyessenov** [comment]
If a process OOMs, good luck cleaning up dangling resources... Restarting pod is a safe procedure, since the whole sandbox goes away.

**Comment by mandarjog** [comment]
@costinm can you attach logs? what configuration is mixer running with?

**Comment by douglas-reid** [comment]
I don't believe that "restartPolicy: Never" is valid for Deployments. We need a different approach.

**Comment by costinm** [comment]
Standard mixer - logs are gone after container restart. It is a load test cluster - but same config has ran for 1 month with 0.8. 

So far crashes of pilot or mixer didn't cause the pod to get stuck - it's just a uds socket that needs to be remove (or reverted to tcp - do we have numbers on performance delta ? ). We still want istio components to work on other enviroments - like VMs - where we can't do the equivalent of restarting a pod. 

Safest approach IMO is just use tcp socket for 1.0, and cleanup and test more for 1.1

**Comment by kyessenov** [comment]
This is a big problem for production users, too. Any application deployment with sidecars can get into this stuck state...

**Comment by kyessenov** [comment]
A standard liveness check is sufficient to fix this (and probably other issues). But we do need to check that proxy correctly handles retries and load balancing between policy pods.

**Comment by costinm** [comment]
Pretty sure no app deployment get stuck in this - it is a problem with the unix socket, and apps don't currently use that. 

Just reverting to TCP socket or cleaning the unix socket will fix this - and will work for VMs too.

**Comment by mandarjog** [comment]
@kyessenov either you can revert to TCP or add liveness/readiness.
Reverting to TCP is a simpler fix.

That said, Mixer not using liveness / readiness is already an issue, so that should be fixed in 1.0 anyway.

**Comment by mandarjog** [comment]
I will add tcp back with uds option.
Can you use build from #7110 , I expect that I won’t get stuck.

More importantly though need to figure out why it is oom.

**Comment by kyessenov** [comment]
@costinm can you provide a profile of mixer when it OOMs? There is a pprof available. Let's focus on that, as UDS vs TCP is a triviality here.

**Comment by costinm** [comment]
Yes, please focus on load testing and making sure it doesn't crash.

You can use test/helm directory to reproduce - or fortio in any setup you
want, I'm running the cluster with around 3000 qps.

Last time it crashed memory was 20G - and I doubt pprof would respond, but
maybe you can catch it earlier.
It is a clear regression and it's far more urgent than the restart issue.

**Comment by costinm** [comment]
You can check 'weekly10' cluster under istio-testing - that's where I run
the load tests.

**Comment by douglas-reid** [comment]
closing, as resolved in UDS fix #7110.
